.github/workflows/branches-and-prs.main.kts

```diff
@@ -0,0 +1,153 @@
+#!/usr/bin/env kotlin
+
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+@file:DependsOn("io.github.typesafegithub:github-workflows-kt:0.40.1")
```

Comment:
Why not add this to the `kotlinScriptClasspath` instead of duplicating it in every `.kts`?

Reply:
A Kotlin `.main.kts` script is a standalone utility that can run on its own.
The native way to call them is to have Kotlin installed and just execute the script either explicitly with Kotlin CLI, or if you are using some `sh` derivative by executing the script.

The Gradle task is just a way I came up with to be able to execute the scripts as Gradle tasks.

The `@DependsOn` within the script is needed so that the normal way to run it works, that is for example also used by the consistency checks and also is faster than ramping up Gradle, which I just added as convenient launcher without the need to have Kotlin installed.

Also the IntelliSense in the IDE would not work if the `@DependsOn` were not there as it then would not know about that dependency.

Regarding the duplication, you could actually drag out common things like defining the `SetupBuildEnv` class or also the `@DependsOn` line to a common script that you then include with `@file:Import`. Unfortunately, this approach has still some bugs in Kotlin. For example you then sometimes loose IntelliSense completely and if there is a change in the common script, you get strange non-obvious compilation errors that you can only fix (as far as I know) by deleting the `.main.kts` compilation cache locally. Hence I did not yet use that approach in Spock. But you could have a look at it at https://github.com/Vampire/setup-wsl/tree/master/.github/workflows where I define some common logic in `workflow-with-copyright.main.kts` and live with the consequences.

Reply:
I think that would be an acceptable downside to get rid of the duplication. Given that those file are not frequently modified.
I'd document the caveat and fix in a readme in the workflows folder.

Reply:
Ok, I refactored some things out to a `common.main.kts` and added the readme, also describing the approach and different options to generate the YAML files, together with the caveats.

Reply:
Thinking more about this and the issues with this approach that you documented.

> A Kotlin `.main.kts` script is a standalone utility that can run on its own. The native way to call them is to have Kotlin installed and just execute the script either explicitly with Kotlin CLI, or if you are using some sh derivative by executing the script.

Do we really need the standalone functionality? Personally, I probably wouldn't run it that way and it seems the downsides are pretty heavy. We could replace the consistency check with a gradle task, or just rely on the GH action to validate the state.

Reply:
> it seems the downsides are pretty heavy.

The downsides are not related to being able to run it standalone.
The first two are bugs of the Kotlin IntelliJ plugin which are happening when editing the scripts. How you run them is irrelevant.
But in one of the according issues it was just said that they plan to improve the whole experience, so hopefully this gets fixed in forseeable future.

The third point is a bug in Kotlin itself, either the compiler or the runtime, I don't know. Indeed there would be another mitigation option, especially as they are not run often. You can disable the compilation cache by setting an evironment variable, so for running through Gradle I could just always disable the compilation cache and that would automatically mitigate the problem there. On GHA it is not a problem anyway as it runs on ephemeral environment so there is no compilation cache used anyway.

Reply:
I added it to the Gradle tasks and mentioned it in the mitigation strategies should someone wish to use one of the other ways.